### ISSUE TITLE
Netty branch/module for early access Loom compatibility

### ISSUE THREAD

**Issue by paul-hammant** [netty/netty#8439]
It would be great if work were started for Java fibers (Project Loom).

There's https://github.com/petrbouda/jlhttp-server-on-fibers/ which seems like a benchmarkable comparison for a minimal HTTP server with and without Fibers. There's also https://github.com/forax/loom-fiber showing deliberate experiments with Fibers and continuations.

It'd be great if there were a branch-by-abstraction technique to composing the core jar from multiple bytecode targets in order to not break backwards compability with pre-loom JVMs. What do I mean by that? Well, XStream goes to a lot of effort to make something that fits all JVMs by using activation profiles - https://github.com/x-stream/xstream/blob/master/xstream/pom.xml

**Comment by normanmaurer** [comment]
I think there are currently not plans to do any experiments by the core devs. That said I would be happy to review any code if someone from the community steps up.

**Comment by chengenzhao** [comment]
+1 loom is awesome and projects like vert.x could start working on java version CoroutineVerticle and Fiber based Worker thread pool etc.

**Comment by lukehutch** [comment]
At Oracle Code ONE 2018, Michael Reinhart demonstrated massive improvements in scalability, throughput and latency skew by switching a network server from `java.lang.Thread` to Loom -- up to an order of magnitude improvement across multiple metrics in some cases. This is definitely something worth doing, with massive potential benefits to Netty and Vert.x.

**Comment by normanmaurer** [comment]
@lukehutch as said before its something we are interested in in general but not a high priority atm. Contributions welcome of course

**Comment by mystdeim** [comment]
+1

**Comment by paul-hammant** [comment]
Perhaps we need to set a bounty for this :)

**Comment by lukehutch** [comment]
@normanmaurer I have no idea how much work would be involved in this. Can you or one of the other Netty developers please give a few quick pointers to where to find the threading code in Netty, and what might be involved in switching to a fibers threading model? (No guarantees I will have time to work on this, but I do want to learn about fibers in a deep way, and this could be a good way to accomplish that.)

**Comment by medusar** [comment]
Great proposal !

**Comment by normanmaurer** [comment]
@lukehutch  the threading code would be in the `EventExecutor` / `EventLoop` implementations.  As what exactly would be needed to be changed, I have no idea yet.

**Comment by vincentfree** [comment]
+1

**Comment by Sevenops** [comment]
You can check this repo https://github.com/Sevenops/NettyFiberBenchmarks

**Comment by lukehutch** [comment]
@Sevenops thanks for the link. I'm surprised the benchmarks are worse with fibers, since they should be strictly more efficient. I guess there is still some optimization work to do.

**Comment by Sevenops** [comment]
> @Sevenops thanks for the link. I'm surprised the benchmarks are worse with fibers, since they should be strictly more efficient. I guess there is still some optimization work to do.

Using fibers for event loop isn't correct usage of fibers but it's performing better than normal threads.
MongoDB benchmarks weren't correct so i've deleted them for now.
I got 80k req/sec with scheduling fiber for every read and it's still good because you can write in sync way.
For future i'll add mysql benchmark with correct usage of  fibers.

**Comment by paul-hammant** [comment]
Cool. What would an unloaded backend look like in benchmarks?  Like /endpoints/calculator/1/addTo/1

**Comment by Sevenops** [comment]
I've added mysql benchmark for fiber version i couldn't get jasync-sql working for netty benchmark.
You can use jar in lib folder to run with fibers it's modified version of mariadb java driver.

**Comment by Sevenops** [comment]
I have made changes to mongodb driver to work with fibers and fibers works slightly better in this case.

**Comment by lukehutch** [comment]
@Sevenops Nice work. Can you please provide a description of the changes that have been needed so far to convert these libraries to use fibers?

**Comment by Sevenops** [comment]
> @Sevenops Nice work. Can you please provide a description of the changes that have been needed so far to convert these libraries to use fibers?

I've added them now https://github.com/Sevenops/NettyFiberBenchmarks/tree/master/patches
If you want to make libraries fiber friendly run java loom with  -Djdk.tracePinnedThreads=full option so it'll give error about where is fiber couldn't yield and you can make them fiber friendly.

**Comment by lukehutch** [comment]
Thanks, I took a look at the patches. But then what is needed to have an executor use fibers rather than threads? And why does wrapping a method with `RentrantLock` make it more fiber-friendly? Is that because `RentrantLock` is itself fiber-friendly, whereas the API calls that it wraps are not fiber-friendly?

**Comment by Sevenops** [comment]
Afaik most of the java.util.concurrent is fiber friendly but synchronized isn't and you need to change them. In repo there is a java file called FiberPerTaskExecutor you can use it example :
`new NioEventLoopGroup(8,new FiberPerTaskExecutor(new FiberFactory()));`

**Comment by lukehutch** [comment]
@Sevenops So there's no way for a future JVM version to cause `synchronized` to yield?

**Comment by Sevenops** [comment]
> @Sevenops So there's no way for a future JVM version to cause `synchronized` to yield?

Probably no but i'm not sure you should ask at loom-dev@openjdk.java.net
I believe after loom is ready library creators will update their codes for loom.

**Comment by paul-hammant** [comment]
Early Access Loom Builds for community feedback - https://mail.openjdk.java.net/pipermail/loom-dev/2019-July/000633.html

**Comment by bxqgit** [comment]
Hi all and @normanmaurer in particular!
I have some questions :)

I may be wrong, but as far as I understand, the whole 'Reactive/Event Loop' thing, and *Netty* in particular, was invented as an answer to the *C10K+* problem. It has obvious drawbacks, as all your code now becomes *Async*, with ugly callbacks, stack traces, and therefore hard to maintain and to reason about.

*Go*'s language with goroutines was a solution, now they can write *Sync* code and also handle *C10K+*. So now *Java* comes up with *Loom*, which essentially copies the *Go*'s solution, soon we will have *Fibers* and *Continuations* and will be able to write *Sync* code again.

So the question(s) are:
1. When the *Loom* is released in production, doesn't it make *Netty* kinda obsolete?
2. If I have *Fibers* and *Continuations* in Java, can one write nice *Sync* code and be ok with *C10K+* without *Netty*?
3. Are there any advatages, in performance or *C10K+*, in writing *Async* code and using *Netty*, after production release of *Loom*?

I understand that *Netty* is more than just 'Reactive/Event Loop' framework, it also has all the codecs for various protocols, which implementations will be useful somehow anyway, even afterwards.

Thanks :)

**Comment by lukehutch** [comment]
My understanding after looking into this (I am not an expert) is that:

1. Most code will port very easily to fibers by just instantiating a fiber pool rather than a thread pool, and nonblocking async code should perform very well with this simple change.
2. Currently object monitors (synchronized locks) cannot yield properly, which can create a scalability issue, but none of Netty should use them anyway, due to its deeply async design. I don't see any reason why it should not be possible for monitors to properly yield with the fibers model eventually, but getting this working has not been straightforward for some reason (anyone know why?).ft
3. You can already use Netty with proper Go-like async/await semantics, a through Kotlin coroutines, eg.: https://vertx.io/docs/vertx-lang-kotlin-coroutines/kotlin
4. I expect Java will get an async/await formalism too once Loom is working, it is inevitable. But until then, you can already add it through bytecode processing: https://github.com/electronicarts/ea-async

**Comment by bxqgit** [comment]
@lukehutch Thanks for your opinion. Though you've answered to a different set of questions: "Can I use *Loom* right now in September 2019?", and "Will I be able to use my *Async* code and *Netty*, but also use *Fibers* and *Continuations* advatages?" , and "What can I do right now if I want to write *Sync* code that works with *C10K+*?" :)

Right now, I'm using *Netty* and my *Async* code works just fine, I don't need any other solutions at the moment.

I was talking about some moment in the future, when *Loom* is released in production JDK, all the problems are solved, and everything in Java Standard Library works and interoperates with *Fibers* and *Continuations* perfectly.

Say it's year 2021 :) And I want to start a new project with *C10K* requirement. Are there any reasons to pick *Netty* and write *Async* code in year 2021? Or should I rather just write *Sync* code, using *Fibers* and *Continuations*, and probably some other lightweight encoding/decoding library for a high-level protocol, like HTTP/Next?

Is *Netty* and it's *Async* architecture going to be obsolete in the near future? Or will it continue to have some advantages even after release of *Loom* and it's *Sync* code model?

**Comment by lukehutch** [comment]
I think I jumped ahead and answered the questions in your second message rather than the first :) It is not possible under Loom to magically turn sync code into async code. You need something like async/await to do that. Java does not support this, but it probably will *after* Loom is available (async/await is a non-goal for Loom currently). But Kotlin supports async/await on the JVM already, today. So use Kotlin coroutines if you want this today, not in 2021.

As far as scaling to the C10K requirement, Netty is particularly well suited to this under Loom, because of its fully asynchronous design. Javascript programmers will tell you one thread is enough for massive scalability under an async nonblocking paradigm -- but they only say that because Javascript has only one thread available without jumping through hoops (yes, this is changing with the introduction of shared memory and native wasm threads). Writing async nonblocking code on a fully mutithteaded system is more scalable, of course, and switching from threads to low-latency fibers is not only still more scalable, but has dramatically lower (and more predictable) latency. So Netty, running under Loom, will perform better than it ever has. And of course you can write your own network stack if you want, but if not, then yes, you will still need Netty.

**Comment by bxqgit** [comment]
@lukehutch 
I think that the subject we are talking about is enough difficult and it looks to me that we really differently understand how things work :)

So, as far as I understand:

0. **Sync** code is better than **Async**. **Reactive/Event Loop** model is complicated and is kinda ad-hoc solution to the **C10K+** problem. My personal preference would be to write plain straight simple **Sync** code, without callbacks, futures, promises, async/await syntax.

1. **Async/Await** is a syntax sugar for using **Futures/Promises**. Which are it turn the solution for **Callback Hell**. All of this is related to **Reactive/Event Loop** model and **Async** code, and has nothing to do with **Loom** and **Sync** code.

2. **Loom** enables **Sync** code to be usable for **C10K+** problems. If you're familiar with **Go** and **Goroutines**, I see **Loom** as almost exact copy. It gives you **Fibers**, as a lightweight replacement of **Threads**. So you can start 100'000s of **Fibers** to handle **C10K** cases.

3. With **Loom** you'll be able to write **Sync** code, which will work just fine in **C10K+** cases, won't consume a lot of memory, won't suffer from context switches, etc. In general, it will have all the advantages which **Go** and **Goroutines** now have in 2019.

4. Here's an example of **Sync** pseudo-code I'm talking about. Notice that there're no callbacks, event loops, futures, promises or async/await syntax:

```
var serverSocket = new ServerSocket(80);
whilte(true) {
  var s = serverSocket.accept();
  (new Fiber() {
    void run() {
      var reqBytes = s.readBytes();
      var resBytes = someProcessing(reqBytes);
      s.writeBytes(resBytes);
    }
  }).start();
}
```

5. With **Loom** the above code scales for **C10K+** cases. But today, we don't have **Loom** and if we switch **Fiber** to **Thread** - everything will still work, but will scale only for 100+ simultaneous users maximum, not a lot.

6. And yes, probably you can add **Loom** support to **Netty**, and use **Fibers** and **Continuations** ... but why?? If you use **Netty**, anyway you're kinda forced into **Reactive/Event Loop** model and **Async** code, which has all the cons described above.

**Comment by lukehutch** [comment]
The exam code snippet you gave *is* async/await code. You apparently just don't have to use the keywords for async/await in Go, which is of course ideal.

Read up on CPS (Continuation Passing Style). This is semantically what is going on under the hood of async/await, although with fibers, this can be accomplished using a yield (which is really the same thing as cooperative multitasking) -- the continuations are implicitly saved on yield rather than code-generated as with CPS.

As I mentioned, it is not possible under Loom to magically turn sync code into async code. At the very minimum, the object monitor lock needs to be able to yield in order to enable this, and (last I checked) this is not yet a solved problem in the JVM.

If you didn't click the link for the Java and Kotlin async/await mechanisms, take a look at the examples -- they are pretty much exactly like your Go code, albeit with some additional (admittedly pesky) keyword/annotation requirements to mark blocking functions as such. I too wish the compiler could just figure that out by itself.

**Comment by bxqgit** [comment]
@lukehutch 

1. The code I've written is not **Go**. It's **Java** using **Loom**.
2. I'm not sure what is 'async/await code without async/await keywords' you're talking about.
3. I think the correct term to talk about would be **Green threads**.
4. **Loom** brings **Green threads** to **Java**.
5. **Go** has **Green threads** as **goroutines**.
6. There're at least **two** approaches of bringing **Green threads** to a language. First is to use 'async/await' constructs, JavaScript, C#, Kotlin use this approach. Second is not to use any additional constructs, **Go** and **Java** with **Loom** use this approach.
7. The syntax of **Java** with  **Loom** won't have any 'async/await' stuff.
8. Compatability with **synchronized**, **locks**, **volatile**, **atomics**, etc will be resolved in production release of **Loom**.
9. **Thread** and **Fiber** classes will be interchangeable, and have a common parent **Strand**.

If you disagree with any of my statements, feel free to comment with a quote. I think that would be a better way to find the answer to the question: Is **Netty** doomed or not? :)

**Comment by chengenzhao** [comment]
‘async/await code without async/await keywords’
Is it some thing like Vert.x 4 doing?
from 
`await webClient.send();`
to 
`webClient.send().await();`
which doesn't require user to use await and async key words

**Comment by chengenzhao** [comment]
Kotlin is using coroutine scope way
Like in Vert.x
```
scope extends CoroutineScope{//CoroutineScope usually is CoroutineVerticle
  launch{
    //no need to async/await key words here
    val result = webClient.send().await();
  }
}
```
For Loom
```
scope extends ContinuationScope{
  //compared to Kotlin, launch statement is even redundant here
  var result = webClient.send().await();
}
```
The default dispatcher will dispatch the callback function to original thread which is eventloop in Vert.x, but still, we need await function even we don't use await key word

but also, Loom allows users to use Fork&Join Pool
Fiber == Continuation + Scheduler(dispatch to Fork&Join Pool)
so Vert.x could provide Fiber based worker verticle
and then in worker verticle
```
scope extends FiberScope{
  var result = webClient.send();//no need to use .await() function to return the value
}
```
but carrier thread maybe changed

So in a nutshell, Loom is a composition of async/await+eventloop and fiber(carrier thread is fork&join pool)
depends on how you implement the scheduler
if the scheduler dispatches the callback to the original thread(ui thread/eventloop), then it is coroutine(async/await)
but if the scheduler dispatches the callback to the fork&join pool, then it is fiber to replace the thread.

Therefor Java&Kotlin etc. are not just use async/await approach, but also Fiber approach(Goroutine) depends on how you define the scheduler.
So Java(Loom), Kotlin(Coroutine) is a superset of Go language and Node(JS) approach
since Go only supports Fiber and Node.js only supports eventloop+async/await
but Java&Kotlin support both

**Comment by lukehutch** [comment]
> If you disagree with any of my statements, feel free to comment with a quote. I think that would be a better way to find the answer to the question: Is **Netty** doomed or not? :)

Not sure if I'm just taking troll bait, but why would Netty be doomed, whether or not it runs better on Loom than threads? It's an important library. If you don't like the API, use something else. If you simply don't like callbacks (I don't either), then as I mentioned earlier, use Kotlin coroutines or EA Async today.

**Comment by bxqgit** [comment]
@chengenzhao 
So, considering what you've just said, does the introduction of **green threads** into **Java** via project **Loom** makes **Netty** obsolete? If one has **green threads**, why would one write **Async** code using **Netty**, if he could write plain simple **Sync** code?

**Comment by slandelle** [comment]
@bxqgit No it doesn't. Green threads will help implement different and more efficient APIs than what's currently possible in Java. And you can be sure Netty will provide one once Loom is production ready.
But you'll still need an underlying non-blocking IO framework that plugs efficiently on OS primitives such as epoll, period.

**Comment by bxqgit** [comment]
@lukehutch 
It's more a philosophical question than a practical one. Looks like you don't follow the whole chain of logical conclusions. You continue argue with individual statements, but don't see the whole picture of why **Netty** may be doomed and become obsolete in the near future, considering emergence of project **Loom**.

Kotlin's coroutines and other async/await libraries have all the downsides of code using **Netty**, In short - it's too complicated. Today, Ocober 2019, I'd use **Go** and its gorotines. But I look forward for **Loom** to bring **green threads** to **JVM** and allow everyone to write nice **Sync** and scalable **Java** code again.

**Comment by bxqgit** [comment]
@slandelle 
Looks to me that OS integration and interaction with **epoll/kqueue** - all of that is one of the main parts of project **Loom**. You will be able to write **Sync** code within **Fibers**, and JVM will use epoll/kqueue and **Async** **IO** underneath. Just like now you can do this with **Goroutines**.

**Comment by bxqgit** [comment]
> Green threads will help implement different and more efficient APIs than what's currently possible in Java. And you can be sure Netty will provide one once Loom is production ready.
Yeah, this probably mean a completely new API, fully **Sync**, nothing in common with the current one. Looks more like a new project, **Loometty** ;)

**Comment by mosesn** [comment]
@bxqgit I'm not sure what you're trying to achieve here–it feels like you've derailed a useful issue thread.  Most of the hard work of Netty is in protocol integrations, and making network stuff work properly.  The Promise API is an implementation detail, compared to the rest of what's in Netty.

**Comment by bxqgit** [comment]
@mosesn

> Most of the hard work of Netty is in protocol integrations, and making network stuff work properly.

I would say **protocol encoders/decoders** are **50%**, and the other **50%** is **Async (Promise) API/Architecture**' to handle **C10K+** problem. And those **protocol encoders/decoders** are kinda locked on the **Async (Promise) API/Architecture**. So it still looks that a complete rewrite in required to leverage all of the features of **Loom**, to be able to write simple **Sync** code... **Loometty 1.0** :)

Anyway, I agree, some **Netty** slack or chat would be a better place for this kind of discussion. But I haven't found any! And this thread looked like an suitable place to try to ask my questions :)

Bye, developers! :)

**Comment by lukehutch** [comment]
Again a rewrite is absolutely not necessary for Netty to make use of Loom, and no, Loom does not make Netty, its architecture or its async model obsolete -- Loom will just make Netty dramatically more scalable and efficient. The simplest way to use Loom with Netty is to schedule Netty threads from a fiber pool rather than a thread pool, which can be as simple as a one-line change, though may require a bit more work due to Netty's complexity.

One day Java will get the ability to call async code in a synchronous-looking way, and this will be able to be used for Netty or anything else using the promises or futures model -- but changing the language syntax is currently a non-goal for the initial release of Loom.

But I am repeating myself for the second or third time now. I hope you had your questions answered.

**Comment by Sevenops** [comment]
> I think there are currently not plans to do any experiments by the core devs. That said I would be happy to review any code if someone from the community steps up.

I'd like to start loom branch is there any advice to how should i implement it ?

**Comment by paul-hammant** [comment]
@Sevenops - https://github.com/lucav76/Fibry doesn't say anything about HTTP, but it is a super solid and buildable/testable demo of Loom for us non Oracle folks.

**Comment by landon30** [comment]
1. [Loom Early-Access](http://jdk.java.net/loom/) build on jdk15. and i guess the GA release will be in jdk17,2021,next jdk lts.
2. we can try kotlin coroutine with vertx.
3. loom is the ulimate concurrency solution

**Comment by Jotschi** [comment]
> > Additionally I want to point out that JNI calls will still block the CarrierThreads and this may cause problems since Netty uses JNI in a few places.
> 
> Netty uses non-blocking IO, but expects to be able to block event loop threads while waiting for IO completions, or for new tasks to be posted. For epoll, kqueue, and io_uring, this blocking will be done in JNI calls.

I'm not 100% sure about this but I believe there might be a limitation with Virtual Threads where the Thread gets pinned as soon as there is a JNI operation on the stack of the carrier thread which executes the operation. I tested also io_uring with virtual threads and more threads got pinned. But I should stress that it is a bit early to come to conclusions since Loom is still in development and may not be included in march in JDK 18 after all.

I don't think I will continue to investigate Loom support for Netty at the moment. Imho Netty can benefit from Virtual Threads by utilizing them in the request handlers. For Vert.x I'm currently [experimenting with code generation](https://github.com/jotschi/vertx-loom) of a wrapper API enables transparent usage of virtual threads.

**Comment by joeyleeeeeee97** [comment]
@normanmaurer @franz1981 @chrisvest 
Any updates on this? 

I believe io_uring with virtual thread(blocking read/write each thread) will bring the best performance.

I did some experiments before #12348 with virtual thread selector , and there are some visible improvements.

**Comment by valerauko** [comment]
Java 21 is out now with GA virtual threads, so I think the priority of this should rise?

**Comment by He-Pin** [comment]
I can submit a pr if this is acceptable @normanmaurer

**Comment by chrisvest** [comment]
Event loop threads can't be virtual without a way to tell the runtime that we're doing a blocking JNI call and it should somehow cope in a way that doesn't deadlock carrier threads.

We could instead offload pipeline processing of messages to a virtual-thread-per-channel. This would incur some synchronization overhead on every read and flush, but the EV threads can then focus on IO, and maybe scheduled tasks.
I imagine this is what people are after, with virtual threads. This will be easier to add in 4.2, with the IoHandle family of APIs we've added there.

**Comment by He-Pin** [comment]
Another issue is that when using virtual thread, the user should go with the `AdaptiveByteBufAllocator` without heavily using a thread local.

**Comment by franz1981** [comment]
Actually there is a other way to integrate with Netty and is to use a the Netty executor to run single carrier virtual threads (Scylladb does something similar with Seastar and coroutines), but requires something on OpenJDK to be opened...stay tuned :)

**Comment by He-Pin** [comment]
I do encounter dead lock when make use of Virtual thread with Netty, where Netty's eventloop and my virtual thread both try to loading some Netty's classes, okay, deadlock because of the classloading :(, one way to fix this is always limit the max concurrent virtual thread been created, dead local because of there is no carrier thread to execute the wakeup virtualthraed.

**Comment by franz1981** [comment]
I read something about it in the loom dev list, am I right or is a different issue?

**Comment by He-Pin** [comment]
I posted that on loom-dev, you may have already read it, yes, maybe a wiki page is enough, but sooner or later, a user will encounter deadlock based on virtual thread's current implementation.

**Comment by He-Pin** [comment]
```java
"mtop-http-worker-12" #1640 [4656] daemon prio=5 os_prio=0 cpu=62.02ms elapsed=48074.82s tid=0x00007faae0128f20 nid=4656 waiting for monitor entry  [0x00007fa8218a8000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at java.lang.ClassLoader.loadClass(java.base@21.0.2/ClassLoader.java:573)
        - waiting to lock <0x0000000545a8ccf0> (a java.lang.Object)
        at io.pandora.classloader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:119)
        at java.lang.ClassLoader.loadClass(java.base@21.0.2/ClassLoader.java:526)
        at io.netty.util.ResourceLeakDetector.track0(ResourceLeakDetector.java:274)
        at io.netty.util.ResourceLeakDetector.track(ResourceLeakDetector.java:250)
        at io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer(AbstractByteBufAllocator.java:44)
        at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:404)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
        at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:116)
        at io.netty.handler.codec.http.HttpObjectEncoder.encodeFullHttpMessage(HttpObjectEncoder.java:328)
        at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:165)
        at io.netty.handler.codec.http.HttpClientCodec$Encoder.encode(HttpClientCodec.java:290)
        at io.netty.handler.codec.http.HttpObjectEncoder.write(HttpObjectEncoder.java:100)
```

@franz1981 It's something like this. We can get around this by limiting the virtual threads count <= carrier threads count.

**Comment by franz1981** [comment]
in quarkus we are working on a lock free class loader for this reason :)

**Comment by He-Pin** [comment]
@franz1981 I think you will need some changes in the JVM too

**Comment by oddbjornkvalsund** [comment]
@Jotschi wrote:
> I however noticed during my testing that Netty would no longer accept requests when using a virtual thread pool size of 500+.

I observed the same thing and lowering size of the virtual thread pool to <250 seems to fix the problem. Note that I only observed this problem running on a Linux VM, not on my developer Macbook. I suspect that the 250 number is related to the default pool size of the JVM Fork-Join pool, but this is just a hunch.
